fix: fire updateActive after calcPoints

DanielRuf opened the pull request:
@ncoden this seems to be the case for the other randomly failing unit tests.

Please verify and test it with a few test iterations.

DanielRuf committed:
fire updateActive after calcPoints

ncoden commented on js/foundation.magellan.js:
```diff
@@ -129,7 +133,7 @@ class Magellan extends Plugin {
     SmoothScroll.scrollToLoc(loc, options, function() {
       _this._inTransition = false;
-      _this._updateActive();
+      //_this._updateActive();
```
No commented code please, never ;)

DanielRuf replied:
Ok.

ncoden commented on js/foundation.magellan.js:
```diff
@@ -100,11 +100,15 @@ class Magellan extends Plugin {
         var arrival   = this.getAttribute('href');
         _this.scrollToLoc(arrival);
+        _this.calcPoints();
+        _this._updateActive();
```
I investigated on this issue before. The easy solution is to put `this.calcPoints();` everywhere like @kball did 1 year ago (https://github.com/zurb/foundation-sites/pull/9491), and according to me this should work.

But `calcPoints()` is a huge operation that will get the position of all targets on the page. When the plugin was implemented, `calcPoints()` was only called on init/reflow so `_updateActive` on the scroll event works on pre-calculated positions. Calling `calcPoints` everywhere is against this optimized pattern.

So I think we should rather do the following:
* Consider a "page is loaded" when we receive the `onload` event OR the page is already loaded when the plugin is instantiated (related to https://github.com/zurb/foundation-sites/pull/10947).
* Wait for the page to load to listen for scroll events (or others events that trigger `_updateActive`).
  If the user begin to scroll before, we cannot update the magellan anchors anyway.
* Wait for the page to load before scrolling in tests.

DanielRuf replied:
Mh, sounds like a small refactor for this. Should we handle it in this PR or another one?

ncoden replied:
I think we should clean that in this PR, as this is "anti-pattern" and can introduce some performance issues.

DanielRuf replied:
Could need some advice. Do you have some specific solution in your mind how this should look like?

ncoden replied:
hmm... For example
1. An an utility to check for `readyState === "complete"`, and will create a listener on `$(window).once('load', ...)` otherwise.
2. Use it in the `_event` function and start listening all others events in it (so we listen nothing before the page is loaded).
3. For the tests, I guess we can still use something like `$html.once('load',  ...)` and call `plugin.scrollToLoc(...)` within.

ncoden reviewed (changes requested):
Nice catch but I don't think that's the best solution.

DanielRuf committed:
remove unused code

DanielRuf committed:
do less calculations to improve the performance

DanielRuf commented:
Pushed a new commit. Please review and test on your side (~80-100 test runs).

ncoden commented on js/foundation.magellan.js:
```diff
@@ -93,15 +93,28 @@ class Magellan extends Plugin {
+          _this.scrollToLoc(arrival);
+        });
+      });
+    }
```
Duplicated content.

We could do something like:
```js
// core.util.js
function onLoad(cb) {
  (document.readyState === "complete")
    ? cb()
    : $(window).one(cb);
 }
```
```js
// foundation.magellan.js
onLoad(() => {
  // code here
});
```

DanielRuf replied:
Some util method for all components? Because we have this duplicated part probably in more components and tests where we do the same check.

DanielRuf replied:
Sounds good. A candidate for another PR for the core utils?

ncoden replied:
Ok, let's merge this as this, I will refactor all that in https://github.com/zurb/foundation-sites/pull/10947.

DanielRuf replied:
Makes sense 👍 That should ensure that each PR produces consistent behavior and better overview. 

Personally I am not a big fan of mixing too much and big changes in one PR and separate them. But this is just my opinion.

ncoden replied:
One PR = one action, with little or big changes. We can do a PR across the whole framework to refactor events on initialization, it's not a problem this we do more or less the same thing in all components and we're not lost in changes.

ncoden commented on test/javascript/components/magellan.js:
```diff
@@ -56,7 +56,9 @@ describe('Magellan', function() {
-      plugin.scrollToLoc(target);      
+      $html.one('load', function(){
+        plugin.scrollToLoc(target);
+			});
```
The function here is executed dynamically, far after the test function exited. So the test happends before the scroll begin and therefore does not test it.

ncoden replied:
A solution would be to wait for `$html.one('load'` to execute the test (and everything with it). I wonder if we can use `async await` in tests (for a simpler code)...

DanielRuf replied:
Ah you mean the `done`callback.

ncoden replied:
I forgot the `done` callback but yes.

```js
// .... initialize plugin

// wait for page to load
$html.one('load', function(){
  // scroll
  plugin.scrollToLoc(target);
  
  // wait for the scroll to takes effect
  setTimeout(function() {
    // ... test everything ...
    // and say we are done
    done();
  });
});
```

But as you see, this is a bit nested. It would be simpler with async/await

DanielRuf replied:
`async await` is not supported in all browsers, for example IE11

ncoden replied:
`async await` is supported almost nowhere. My point is that test files are transpiled (see [gulp/tasks/test.js:35](https://github.com/zurb/foundation-sites/blob/develop/gulp/tasks/test.js#L35)). But our current global babel config ([.babelrc](https://github.com/zurb/foundation-sites/blob/DanielRuf/fix/develop-dropdown-eventnames/.babelrc)) does not allow `async/await` (see https://medium.com/@dandv/using-es2016-es7-async-await-with-babel-e758bab1c4d5).

ncoden commented on test/javascript/components/magellan.js:
```diff
@@ -82,7 +84,9 @@ describe('Magellan', function() {
-	      plugin.scrollToLoc(target); 
+        $html.one('load', function(){
+          plugin.scrollToLoc(target);
+        });
```
Same here. No error triggered inside `scrollToLoc` will be triggered here.

See for example:
```js
// Numbers indicate the order in which each line is executed
...                               /* 1 */
try {                            
  setTimeout(/* 2 */ () => {
    throw new Error('err');       /* 7 */
  }, 0);                          /* 3 */
} catch(err) {                    /* [not executed] would be 4 */
  console.log('catched');         /* [not executed] would be 5 */
}
...                               /* 6 */
```

The code above will trigger an error. It will not be catched.

DanielRuf replied:
True. Can you show or provide the solution? Currently not 100% clear to me how I can solve these things here as this component feels so tricky =/

DanielRuf committed:
tests: make sure that the Magellan tests are properly running

DanielRuf commented:
It seems the `calls reflow on viewport size change once` test also fails sometimes.

ncoden commented:
> tests: make sure that the Magellan tests are properly running

hmm It seems to be a rollback. Tests may running but:
* the race condition should be back on `it scrolls the selected element into viewport`
* we do not actually test the plugin behavior on `fails gracefully when target does not exist` as we do not call the method when target does not exist.

DanielRuf commented:
> the race condition should be back on it scrolls the selected element into viewport

Could not verify it at least. Did around 70 - 80 runs and did not see them again.

ncoden committed:
tests: revert to systematic test of magellan graceful failure

Partially revert e89d85cbf4328c8d925a64bde7e2525c88841d5a

Test Magellan.scrollToLoc systematically as it should _never_ throw an error when target does not exists.

ncoden committed:
clean: remove comment about a resolved bug in Magellan tests

https://github.com/zurb/foundation-sites/pull/10906

ncoden commented:
> All checks have passed

🍾

DanielRuf commented:
🎉 ✨

ncoden commented:
Note: there is probably a race condition in Interchange, I got this in 1 test on 100:
> * Interchange events() calls reflow on viewport size change once
>     Uncaught Error: Script error. (:0)

DanielRuf commented:
> Note: there is probably a race condition in Interchange, I got this in 1 test on 100:

https://github.com/zurb/foundation-sites/pull/10988#issuecomment-370040655

DanielRuf commented:
Not yet 100% sure what the cause is.
